userguide/tutorials/payment_plugin.adoc

```diff
@@ -1,61 +1,575 @@
+
+. The client application *completes* the *GoCardless Redirect flow* 
+
+. GoCardless then actually sets up the mandate and returns a *mandate Id*
+
+. The *mandate Id* can then be used to charge the customer
+
+=== Using GoCardless from Kill Bill
+
+In order to use GoCardless from Kill Bill, we will need to create a payment plugin corresponding to GoCardless. Since we are developing a very basic plugin that can only process payments, we only need to do the following:
+
+. Set up the mandate. This is a two step process as explained above where:
+
+.. Step 1 involves redirecting the user to a page to manually to set up the mandate. This can be implemented via the `PaymentPluginApi#buildFormDescriptor` method
+
+.. Step 2 involves completing the GoCardless flow and retrieving the mandate Id. This can be implemented via the `PaymentPluginApi#addPaymentMethod` method
+
+. Implement the `PaymentPluginApi#purchasePayment` method to charge the customer
+
+Thus, the end-to-end flow will be as follows:
+
+. The client application invokes the `GoCardlessPlugin#buildFormDescriptor` method with a *success page URL*
+
+. The `GoCarldessPlugin#buildFormDescriptor` method *initiates* the *GoCardless redirect flow* as explained https://developer.gocardless.com/getting-started/api/adding-your-first-customer/#setting-up-direct-debit-mandates-using-hosted-payment-pages[here] and obtains the *redirect URL* which is returned to the client application
+
+. The client application redirects the user to this URL and the user sets up the mandate manually on this page
+
+. GoCardless redirects the user to the *success page*. This invokes the `GoCarldessPlugin#addPaymentMethod`
+
+. The `GoCarldessPlugin#addPaymentMethod` *completes* the *redirect flow* as explained https://developer.gocardless.com/getting-started/api/adding-your-first-customer/#setting-up-direct-debit-mandates-using-hosted-payment-pages[here] and obtains the *mandate id* which is saved in the Kill Bill database
+
+. The client application can then invoke `GoCarldessPlugin#purchasePayment`. This charges the customer against the saved *mandate id*  as explained https://developer.gocardless.com/getting-started/api/taking-your-first-payment/[here]
+
+=== Creating the GoCardless Plugin
+
+Let us now understand how we can create a payment plugin for GoCardless.
+
+==== Step 1 - Initial Setup
+
+The first step in creating a payment plugin would be to create a new Maven project. For this we need to do the following:
+
+. Create a new Maven project
+
+. Copy the https://github.com/killbill/killbill-hello-world-java-plugin/blob/719779b1ea25c14928e92996b6aa21cc9bf8d4fe/pom.xml[Hello World Plugin pom.xml]
+
+. Add the following additional Maven dependency (This is required to access the https://github.com/killbill/killbill-commons/blob/aa83708f56377aabff8391c3ddc197817ad19ec2/clock/src/main/java/org/killbill/clock/Clock.java[org.killbill.clock.Clock] class):
```

Comment:
Maybe we should just add the clock dependency to the hello-world plugin pom.xml instead, that way we can omit this step (it could be confusing for the reader)?

If you agree, can you sent a PR?

Reply:
@Pierre, if the clock dependency is added to the **hello-world plugin** `pom.xml`, the build fails because of unused dependencies. So, it causes the following errors:

```
[WARNING] Unused declared dependencies found:
[WARNING]    org.kill-bill.commons:killbill-clock:jar:0.24.6:compile
ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:3.1.2:analyze-only (default) on project hello-world-plugin: Dependency problems found -> [Help 1]

```
So, the hello-world-plugin needs to be compiled as follows to ignore this error:

`mvn clean install -Dcheck.fail-dependency=false`

Would that be okay? We would also need to update the [plugin development document build instructions](https://docs.killbill.io/latest/plugin_development.html#_build)

Reply:
Yeah, let's simplify the learning for the reader and relax the properties in the pom.xml directly: https://github.com/killbill/killbill-hello-world-java-plugin/blob/719779b1ea25c14928e92996b6aa21cc9bf8d4fe/pom.xml#L43

```
        <check.skip-dependency>true</check.skip-dependency>
        <check.skip-dependency-scope>true</check.skip-dependency-scope>
        <check.skip-dependency-versions>true</check.skip-dependency-versions>
        <check.skip-duplicate-finder>true</check.skip-duplicate-finder>
        <check.skip-enforcer>true</check.skip-enforcer>
        <check.skip-rat>true</check.skip-rat>
        <check.skip-spotbugs>true</check.skip-spotbugs>
```

The hello world plugin doesn't need to be production-ready / perfect, just easy to use :-)

Reply:
@pierre , thanks, I did not realize that properties could be set in the `pom` file :). I've created a PR - https://github.com/killbill/killbill-hello-world-java-plugin/pull/7